File: CSharpPools/PoolsLibrary/ObjectUtilization/ObjectUtilizer.cs
using System;

namespace PoolsLibrary.ObjectUtilization
{
    public class ObjectUtilizer<TK, TV> : IObjectUtilizer<TK, TV>
    {
        public event EventHandler<GoneObjectEventArgs<TK, TV>> ObjectIsGone;

        public void Utilize(TK key, TV pooObject, object caller)
        {
            if (ObjectIsGone != null)
                ObjectIsGone(this, new GoneObjectEventArgs<TK, TV>
                {
                    Key = key,
                    PoolObject = pooObject,
                    Reporter = caller,
                });
        }
    }
}

File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/objectutilization/ObjectUtilizer.java
package ru.urfu.javapools.poolslibrary.objectutilization;

import java.util.HashSet;

public class ObjectUtilizer<TK,TV> implements IObjectUtilizer<TK,TV> {

	private final HashSet<IObjectUtilizationListener<TK,TV>> _listeners;
	
	public ObjectUtilizer () {
		_listeners = new HashSet<IObjectUtilizationListener<TK,TV>>();
	}
	
	@Override
	public void utilize(TK key, TV pooObject, Object caller) {
		
		GoneObjectEvent<TK,TV> event = new GoneObjectEvent<TK,TV>() {{
			setReporter(caller);
			setKey(key);
			setPoolObject(pooObject);
		}};
		
		for (IObjectUtilizationListener<TK,TV> listener : _listeners)
			listener.onObjectUtilization(event);			
	}

	@Override
	public void addListener(IObjectUtilizationListener<TK, TV> listener) {
		
		if (listener != null)
			_listeners.add(listener);		
	}
	
	@Override
	public void removeListener(IObjectUtilizationListener<TK, TV> listener) {
		
		if (listener != null)
			_listeners.remove(listener);
	}	

	@Override
	public int getListenersCount() {
		return _listeners.size();
	}
}